docs/components/zeebe/deployment-guide/interceptors.md

```diff
@@ -0,0 +1,172 @@
+
+> This functionality is currently only available in Camunda Cloud Self-Managed.
+
+All communication from a client to a broker must first pass through a gateway.
+There they can be intercepted before being dispatched. Zeebe provides a way to
+load arbitrary interceptors into the gateway. Some typical examples of what you
+can accomplish with this include:
+
+- enforcing custom authorization rules on incoming calls
+- monitoring and logging of incoming calls
+
+## Implementing an interceptor
+For the communication between client and gateway, Zeebe uses the gRPC
+[protocol](/components/zeebe/technical-concepts/protocols.md). An interceptor is
+thus implemented as a gRPC
+[ServerInterceptor](https://grpc.github.io/grpc-java/javadoc/io/grpc/ServerInterceptor.html).
+
+An implementation must adhere to the following requirements:
+- it implements
+  [ServerInterceptor](https://grpc.github.io/grpc-java/javadoc/io/grpc/ServerInterceptor.html)
+- it has public visibility
+- it has a public default constructor (i.e. no-arg constructor)
+
+Let's consider an interceptor that provides logging of incoming calls as an
+example. Other ServerInterceptor examples can be found in the official grpc-java
+[examples](https://github.com/grpc/grpc-java/tree/v1.41.0/examples).
+
+```java
+import io.grpc.ForwardingServerCallListener.SimpleForwardingServerCallListener;
+import io.grpc.Metadata;
+import io.grpc.ServerCall;
+import io.grpc.ServerCall.Listener;
+import io.grpc.ServerCallHandler;
+import io.grpc.ServerInterceptor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A simple interceptor that logs each incoming call. The class must be public
+ * since we will load it via JAR into the gateway.
+ */
+public final class LoggingInterceptor implements ServerInterceptor {
+  private static final Logger LOGGER =
+      LoggerFactory.getLogger("LoggingInterceptor");
+
+  @Override
+  public <ReqT, RespT> Listener<ReqT> interceptCall(
+      final ServerCall<ReqT, RespT> call,
+      final Metadata headers,
+      final ServerCallHandler<ReqT, RespT> next) {
+    final var listener = next.startCall(call, headers);
+    return new SimpleForwardingServerCallListener<>(listener) {
+      @Override
+      public void onMessage(final ReqT message) {
+        LOGGER.trace("intercepted a call");
+        super.onMessage(message);
+      }
+    };
+  }
+}
+```
+
+## Packaging an interceptor
+Next, you need to package the interceptor class into a fat JAR. Such a JAR must
+contain all classes (i.e. including all classes your own classes depend upon at
+runtime).
+
+There are many ways to do this, but for simplicity we'll use `javac` and `jar`
+directly. Note, that means we have to define a java manifest file by hand, in
+order to place the libraries' classes on the classpath.
+
+```sh
+# to compile LoggingInterceptor.java, we'll need to provide the grpc and slf4j api libraries
+javac -classpath .:lib/grpc-api.jar:lib/slf4j-api.jar ./LoggingInterceptor.java
```

Comment:
:wrench: You don't need to add grpc-api and slf4j-api. These are provided by the gateway itself. The only time you need to provide them is if you're using a different version - however since we'll load the server interceptor from Zeebe, you probably want to fix grpc-api at least to always be the same version as Zeebe's. I'm not sure how to word all this in an easy to understand way, though - the idea is you only need to add libraries which are not already provided by the gateway, _but_ that's risky and makes you rely on them. I guess the ideal would be for us to have a list of dependencies which we guarantee are always provided - SLF4J, grpc, etc. ? Let me know what you think. I think it's also not too bad to otherwise tell users to always package everything, and add a note for advanced users who know what they're doing and explain how the JAR class loading work - you first look up a class in the JAR itself, and if it's not found, fallback to the Zeebe application's classloader. So some dependencies may be omitted at your own risk to reduce JAR file size.

Reply:
In order to compile this class, you need to add these api's. Otherwise you'll end up with compile errors.

Considering runtime dependencies, I though that we use a disconnected classloader for the interceptors. So they cannot mess with the zeebe classes. Perhaps I misunderstood that. As I understand it now from your description, our classloader is parent of the interceptors classloader, if a class in their loader wants to look-up a class defined in our classloader then it can actually reach it. Is that correct?

Coming back to the original example, I feel like an exhaustive list of dependencies we guarantee are available is a bit crazy. I'd rather we tell them to provide all runtime dependencies themselves and provide an actual fat JAR. That should work in all cases and is easiest to explain